exonum/src/storage/keys.rs

```diff
@@ -227,6 +229,26 @@ impl StorageKey for String {
     }
 }
 
+impl StorageKey for SystemTime {
+    fn size(&self) -> usize {
+        12
+    }
+
+    fn write(&self, buffer: &mut [u8]) {
+        let duration = self.duration_since(UNIX_EPOCH).unwrap();
+        let secs = duration.as_secs();
+        let nanos = duration.subsec_nanos();
+        secs.write(&mut buffer[0..8]);
+        nanos.write(&mut buffer[8..12]);
+    }
+
+    fn read(buffer: &[u8]) -> Self {
+        let secs = u64::read(&buffer[0..8]);
+        let nanos = u32::read(&buffer[8..12]);
```

Comment:
I think `BigEndian::read_u64` would be slightly more clear here.

Reply:
I am not sure, but I think we must check that naonos <= 10**9? (at least a debug_assert is required I think).

Reply:
Yeah, if `SystemTime` can be received from an external source, we need to do *something* with the fact that `nanos` can be greater than `1e9`. If left unchecked, it could break canonicity property of serialization:

```rust
let t = UNIX_EPOCH + Duration::new(0, 1_000_000_001);
let other_t = UNIX_EPOCH + Duration::new(1, 1);
assert_eq!(t, other_t);
```

Reply:
> I think BigEndian::read_u64 would be slightly more clear here.

Not sure. In a hypothetical situation when we decide to store everything using `LittleEndian` it will be one more place that should be fixed. I prefer to reuse the implementation.

Reply:
@slowli As far as I can see, `Duration` [performs normalization internally](https://doc.rust-lang.org/std/time/struct.Duration.html#method.new): 
> If the number of nanoseconds is greater than 1 billion (the number of nanoseconds in a second), then it will carry over into the seconds provided.

Reply:
@DarkEld3r Yeah, I know, this is what my example is about 🙂 My problem is that `(0u64, 1_000_000_001u32)` and `(1u64, 1u32)` encode exactly the same `SystemTime` while having different binary presentations (which, afaik, is our definition of non-canonicity). That is, the roundtrip with `SystemTime` does not necessarily work, at the very least.

cc @vldm @DenisKolodin

Reply:
@slowli I believe that the current code is correct, because it is assumed that bytes inside the storage are trusted (I am not sure about the validity of this assumption, but this is how the code works today). We don't validate messages we read from storage either. 

However, I am all for adding an assertion here, for the purposes of documenting the assumption (and for a 100% impossible case of a bug somewhere, which violates the assumption :) )

Reply:
I've mistaken `StorageKey` / `StorageValue` with message-related stuff; so the concerns regarding canonicity should be addressed to it. However, there is one more question regarding ordering of `SystemTime` in `StorageKey`s: without checking `nanos` bounds, it would lead to incorrect ordering in some cases (e.g., `(0u64, 2_000_000_001u32)` is greater than `(1u64, 0u32)`). So it looks like the assertion is a must.